File: native-deps/protobuf-shim/klib-patcher/src/main/java/kotlinx/rpc/protobuf/nativedeps/tooling/KlibPatcher.java
/*
 * Copyright 2023-2026 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
 */

package kotlinx.rpc.protobuf.nativedeps.tooling;

import kotlin.metadata.KmAnnotation;
import kotlin.metadata.KmClass;
import kotlin.metadata.KmConstructor;
import kotlin.metadata.KmFunction;
import kotlin.metadata.KmPackage;
import kotlin.metadata.KmProperty;
import kotlin.metadata.KmTypeAlias;
import kotlin.metadata.internal.common.KmModuleFragment;
import kotlinx.metadata.klib.KlibModuleFragmentReadStrategy;
import kotlinx.metadata.klib.KlibModuleFragmentWriteStrategy;
import kotlinx.metadata.klib.KlibModuleMetadata;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

/**
 * Patches the produced protobuf-shim cinterop KLIB after cinterop finishes.
 *
 * <p>protobuf-shim is published as a KLIB because the build needs it, but the declarations generated into the
 * {@code kotlinx.rpc.protobuf.internal.cinterop} package are internal implementation detail, not supported public API. The cinterop
 * toolchain does not load compiler plugin registrars on this build path, so the narrow reliable hook is to rewrite
 * the produced KLIB metadata immediately after cinterop and before the artifact is consumed or published.
 */
public final class KlibPatcher {
    private static final String TARGET_PACKAGE_NAME = "kotlinx.rpc.protobuf.internal.cinterop";
    private static final String INTERNAL_NATIVE_RPC_API_ANNOTATION_CLASS_NAME =
        "kotlinx/rpc/protobuf/internal/InternalNativeRpcApi";
    private static final String INTERNAL_NATIVE_RPC_API_DEPENDENCY_UNIQUE_NAME =
        "org.jetbrains.kotlinx\\:kotlinx-rpc-protobuf-shim-annotation";
    private static final KmAnnotation INTERNAL_NATIVE_RPC_API_ANNOTATION =
        new KmAnnotation(INTERNAL_NATIVE_RPC_API_ANNOTATION_CLASS_NAME, Collections.emptyMap());
    private static final Method FRAGMENT_PACKAGE_NAME_ACCESSOR = createFragmentPackageNameAccessor();

    private KlibPatcher() {
    }

    public static void main(String[] args) {
        if (args.length != 2) {
            throw new IllegalArgumentException("Expected <unpacked-klib-dir> <packed-klib-file>, got " + List.of(args));
        }

        patch(new File(args[0]), new File(args[1]));
    }

    public static void patch(File unpackedKlibDir, File packedKlibFile) {
        Path tempKlibDir = null;
        try {
            tempKlibDir = Files.createTempDirectory("protobuf-shim-internal-native-rpc-api-klib-");
            copyRecursively(unpackedKlibDir.toPath(), tempKlibDir);

            File archiveDir = resolveArchiveDir(tempKlibDir.toFile());
            File linkdataDir = new File(archiveDir, "linkdata");
            if (!linkdataDir.isDirectory()) {
                throw new IllegalStateException("Missing linkdata directory: " + linkdataDir.getAbsolutePath());
            }

            LinkdataLibraryProvider metadataProvider = new LinkdataLibraryProvider(linkdataDir);
            KlibModuleMetadata metadata = KlibModuleMetadata.Companion.read(
                metadataProvider,
                KlibModuleFragmentReadStrategy.Companion.getDEFAULT()
            );
            List<KmModuleFragment> targetFragments = metadata.getFragments()
                .stream()
                .filter(fragment -> TARGET_PACKAGE_NAME.equals(fragmentPackageName(fragment)))
                .collect(Collectors.toList());

            if (targetFragments.isEmpty()) {
                throw new IllegalStateException(
                    "Expected protobuf-shim cinterop metadata for package " + TARGET_PACKAGE_NAME + " in " +
                        unpackedKlibDir.getAbsolutePath()
                );
            }

            // annotate all kotlin fragments (constructors, functions, properties, type aliases) with
            // the @InternalNativeRpcApi annotation
            targetFragments.forEach(KlibPatcher::annotateFragment);

            writeLinkdata(
                linkdataDir,
                metadataProvider,
                metadata.write(KlibModuleFragmentWriteStrategy.Companion.getDEFAULT())
            );

            // add annotation as dependency in the manifest
            patchManifest(new File(archiveDir, "manifest"));
            // zips the new klib dir to the packed klib file
            repackKlib(tempKlibDir, archiveDir.toPath(), packedKlibFile.toPath());
        } catch (IOException exception) {
            throw new RuntimeException("Failed to patch " + packedKlibFile.getAbsolutePath(), exception);
        } finally {
            if (tempKlibDir != null) {
                deleteRecursively(tempKlibDir);
            }
        }
    }

    private static Method createFragmentPackageNameAccessor() {
        try {
            return Class.forName("kotlinx.metadata.klib.KlibModuleMetadataKt")
                .getDeclaredMethod("access$fqNameOrFail", KmModuleFragment.class);
        } catch (ClassNotFoundException | NoSuchMethodException exception) {
            throw new ExceptionInInitializerError(exception);
        }
    }

    private static File resolveArchiveDir(File unpackedKlibDir) {
        if (new File(unpackedKlibDir, "linkdata").isDirectory()) {
            return unpackedKlibDir;
        }
        if (new File(unpackedKlibDir, "default/linkdata").isDirectory()) {
            return new File(unpackedKlibDir, "default");
        }
        return unpackedKlibDir;
    }

    private static void annotateFragment(KmModuleFragment fragment) {
        KmPackage packageFragment = Objects.requireNonNull(
            fragment.getPkg(),
            "Expected package metadata for " + fragmentPackageName(fragment)
        );
        packageFragment.getFunctions().forEach(KlibPatcher::annotate);
        packageFragment.getProperties().forEach(KlibPatcher::annotate);
        packageFragment.getTypeAliases().forEach(KlibPatcher::annotate);
        fragment.getClasses().forEach(KlibPatcher::annotate);
    }

    private static void annotate(KmClass clazz) {
        addInternalNativeRpcApiMarker(clazz.getAnnotations());
        clazz.getConstructors().forEach(KlibPatcher::annotate);
        clazz.getFunctions().forEach(KlibPatcher::annotate);
        clazz.getProperties().forEach(KlibPatcher::annotate);
        clazz.getTypeAliases().forEach(KlibPatcher::annotate);
    }

    private static void annotate(KmConstructor constructor) {
        addInternalNativeRpcApiMarker(constructor.getAnnotations());
    }

    private static void annotate(KmFunction function) {
        addInternalNativeRpcApiMarker(function.getAnnotations());
    }

    private static void annotate(KmProperty property) {
        addInternalNativeRpcApiMarker(property.getAnnotations());
    }

    private static void annotate(KmTypeAlias typeAlias) {
        addInternalNativeRpcApiMarker(typeAlias.getAnnotations());
    }

    private static void addInternalNativeRpcApiMarker(List<KmAnnotation> annotations) {
        boolean present = annotations.stream()
            .anyMatch(annotation -> INTERNAL_NATIVE_RPC_API_ANNOTATION_CLASS_NAME.equals(annotation.getClassName()));
        if (!present) {
            annotations.add(INTERNAL_NATIVE_RPC_API_ANNOTATION);
        }
    }

    private static String fragmentPackageName(KmModuleFragment fragment) {
        try {
            return (String) FRAGMENT_PACKAGE_NAME_ACCESSOR.invoke(null, fragment);
        } catch (IllegalAccessException | InvocationTargetException exception) {
            throw new RuntimeException("Failed to resolve package name for KLIB fragment", exception);
        }
    }

    private static void writeLinkdata(
        File linkdataDir,
        LinkdataLibraryProvider metadataProvider,
        KlibModuleMetadata.SerializedKlibMetadata serializedMetadata
    ) throws IOException {
        Files.write(new File(linkdataDir, "module").toPath(), serializedMetadata.getHeader());

        List<String> fragmentNames = serializedMetadata.getFragmentNames();
        List<List<byte[]>> fragments = serializedMetadata.getFragments();

        for (int index = 0; index < fragmentNames.size(); index++) {
            String packageName = fragmentNames.get(index);
            List<byte[]> packageFragments = fragments.get(index);
            File packageDir = new File(linkdataDir, metadataProvider.packageDirectoryName(packageName));
            List<String> outputNames = metadataProvider.existingPartNames(packageName);
            if (outputNames.size() != packageFragments.size()) {
                outputNames = generatedPartNames(packageName, packageFragments.size());
            }

            deleteRecursively(packageDir.toPath());
            Files.createDirectories(packageDir.toPath());

            for (int partIndex = 0; partIndex < packageFragments.size(); partIndex++) {
                Files.write(new File(packageDir, outputNames.get(partIndex)).toPath(), packageFragments.get(partIndex));
            }
        }
    }

    private static List<String> generatedPartNames(String packageName, int fragmentCount) {
        String suffix = packageName.contains(".")
            ? packageName.substring(packageName.lastIndexOf('.') + 1)
            : packageName;
        if (suffix.isBlank()) {
            suffix = "root";
        }
        suffix = suffix.replace('.', '_');

        List<String> names = new ArrayList<>(fragmentCount);
        for (int index = 0; index < fragmentCount; index++) {
            names.add(index + "_" + suffix + ".knm");
        }
        return names;
    }

    private static void patchManifest(File manifestFile) throws IOException {
        if (!manifestFile.isFile()) {
            throw new IllegalStateException("Missing KLIB manifest: " + manifestFile.getAbsolutePath());
        }

        List<String> updatedLines = new ArrayList<>();
        for (String line : Files.readAllLines(manifestFile.toPath())) {
            if (!line.startsWith("depends=")) {
                updatedLines.add(line);
                continue;
            }

            List<String> dependencies = Stream.of(line.substring("depends=".length()).split(" "))
                .filter(dependency -> !dependency.isBlank())
                .collect(Collectors.toCollection(ArrayList::new));
            if (!dependencies.contains(INTERNAL_NATIVE_RPC_API_DEPENDENCY_UNIQUE_NAME)) {
                dependencies.add(INTERNAL_NATIVE_RPC_API_DEPENDENCY_UNIQUE_NAME);
            }
            updatedLines.add("depends=" + String.join(" ", dependencies));
        }

        Files.writeString(manifestFile.toPath(), String.join("\n", updatedLines) + "\n");
    }

    private static void repackKlib(Path unpackedKlibInputDir, Path archiveDir, Path packedKlibFile) throws IOException {
        Path archiveRoot = Objects.equals(archiveDir.getParent(), unpackedKlibInputDir)
            ? unpackedKlibInputDir
            : Objects.requireNonNull(archiveDir.getParent(), archiveDir + " must have a parent directory");

        Files.createDirectories(Objects.requireNonNull(packedKlibFile.getParent(), "Packed KLIB must have a parent"));

        try (ZipOutputStream zip = new ZipOutputStream(Files.newOutputStream(packedKlibFile))) {
            try (Stream<Path> paths = Files.walk(archiveDir)) {
                for (Path path : paths.filter(Files::isRegularFile).sorted().collect(Collectors.toList())) {
                    String entryName = archiveRoot.relativize(path).toString().replace(File.separatorChar, '/');
                    zip.putNextEntry(new ZipEntry(entryName));
                    Files.copy(path, zip);
                    zip.closeEntry();
                }
            }
        }
    }

    private static void copyRecursively(Path sourceDir, Path targetDir) throws IOException {
        try (Stream<Path> paths = Files.walk(sourceDir)) {
            for (Path sourcePath : paths.collect(Collectors.toList())) {
                Path relative = sourceDir.relativize(sourcePath);
                Path targetPath = targetDir.resolve(relative.toString());
                if (Files.isDirectory(sourcePath)) {
                    Files.createDirectories(targetPath);
                } else {
                    Files.createDirectories(Objects.requireNonNull(targetPath.getParent()));
                    Files.copy(sourcePath, targetPath, StandardCopyOption.REPLACE_EXISTING);
                }
            }
        }
    }

    private static void deleteRecursively(Path path) {
        if (!Files.exists(path)) {
            return;
        }

        try (Stream<Path> paths = Files.walk(path)) {
            for (Path candidate : paths.sorted(Comparator.reverseOrder()).collect(Collectors.toList())) {
                Files.deleteIfExists(candidate);
            }
        } catch (IOException exception) {
            throw new RuntimeException("Failed to delete " + path, exception);
        }
    }

    private static final class LinkdataLibraryProvider implements KlibModuleMetadata.MetadataLibraryProvider {
        private final File linkdataDir;

        private LinkdataLibraryProvider(File linkdataDir) {
            this.linkdataDir = linkdataDir;
        }

        @Override
        public byte[] getModuleHeaderData() {
            try {
                return Files.readAllBytes(new File(linkdataDir, "module").toPath());
            } catch (IOException exception) {
                throw new RuntimeException(exception);
            }
        }

        @Override
        public Set<String> packageMetadataParts(String fqName) {
            return packagePartFiles(fqName, false).stream()
                .map(File::getName)
                .collect(Collectors.toCollection(LinkedHashSet::new));
        }

        @Override
        public byte[] packageMetadata(String fqName, String partName) {
            try {
                return Files.readAllBytes(new File(new File(linkdataDir, packageDirectoryName(fqName)), partName).toPath());
            } catch (IOException exception) {
                throw new RuntimeException(exception);
            }
        }

        private List<String> existingPartNames(String packageName) {
            return packagePartFiles(packageName, true).stream().map(File::getName).collect(Collectors.toList());
        }

        private String packageDirectoryName(String packageName) {
            return "package_" + packageName;
        }

        private List<File> packagePartFiles(String packageName, boolean requireDirectory) {
            File packageDir = new File(linkdataDir, packageDirectoryName(packageName));
            if (!packageDir.isDirectory()) {
                if (requireDirectory) {
                    throw new IllegalStateException(
                        "Missing package metadata directory for " + packageName + ": " + packageDir.getAbsolutePath()
                    );
                }
                return List.of();
            }

            File[] files = packageDir.listFiles(File::isFile);
            if (files == null) {
                return List.of();
            }

            return Stream.of(files)
                .sorted(Comparator.comparing(File::getName))
                .collect(Collectors.toList());
        }
    }
}

File: native-deps/grpc-shim/klib-patcher/src/main/kotlin/kotlinx/rpc/grpc/nativedeps/tooling/KlibPatcher.kt
/*
 * Copyright 2023-2026 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
 */

@file:OptIn(kotlin.metadata.ExperimentalAnnotationsInMetadata::class)

package kotlinx.rpc.grpc.nativedeps.tooling

import kotlin.metadata.KmAnnotation
import kotlin.metadata.KmClass
import kotlin.metadata.KmConstructor
import kotlin.metadata.KmFunction
import kotlin.metadata.KmProperty
import kotlin.metadata.KmTypeAlias
import kotlin.metadata.internal.common.KmModuleFragment
import kotlinx.metadata.klib.KlibModuleMetadata
import java.io.File
import java.lang.reflect.Method
import java.nio.file.Files
import kotlin.io.path.createTempDirectory
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream

/**
 * Patches the produced grpc-shim cinterop KLIB after cinterop finishes.
 *
 * grpc-shim is published as a KLIB because the build needs it, but the declarations generated into the
 * `grpcCoreInterop` package are internal implementation detail, not supported public API. The cinterop toolchain
 * does not load compiler plugin registrars on this build path, so the narrow reliable hook is to rewrite the
 * produced KLIB metadata immediately after cinterop and before the artifact is consumed or published.
 */
internal object KlibPatcher {
    private const val targetPackageName = "grpcCoreInterop"
    private const val internalNativeRpcApiAnnotationClassName = "kotlinx/rpc/grpc/nativedeps/InternalNativeRpcApi"
    private const val internalNativeRpcApiDependencyUniqueName =
        "org.jetbrains.kotlinx\\:kotlinx-rpc-grpc-core-shim-annotation"
    private val internalNativeRpcApiAnnotation = KmAnnotation(internalNativeRpcApiAnnotationClassName, emptyMap())
    private val fragmentPackageNameAccessor: Method = Class.forName("kotlinx.metadata.klib.KlibModuleMetadataKt")
        .getDeclaredMethod("access\$fqNameOrFail", KmModuleFragment::class.java)

    @JvmStatic
    fun main(args: Array<String>) {
        require(args.size == 2) {
            "Expected <unpacked-klib-dir> <packed-klib-file>, got ${args.toList()}"
        }
        patch(
            unpackedKlibDir = File(args[0]),
            packedKlibFile = File(args[1]),
        )
    }

    internal fun patch(unpackedKlibDir: File, packedKlibFile: File) {
        val tempKlibDir = createTempDirectory(prefix = "grpc-shim-unsupported-api-klib-").toFile()
        try {
            unpackedKlibDir.copyRecursively(target = tempKlibDir, overwrite = true)
            val archiveDir = resolveArchiveDir(tempKlibDir)
            val linkdataDir = archiveDir.resolve("linkdata")
            check(linkdataDir.isDirectory) {
                "Missing linkdata directory: ${linkdataDir.absolutePath}"
            }

            val metadataProvider = LinkdataLibraryProvider(linkdataDir)
            val metadata = KlibModuleMetadata.read(metadataProvider)
            val targetFragments = metadata.fragments.filter { fragmentPackageName(it) == targetPackageName }

            check(targetFragments.isNotEmpty()) {
                "Expected grpc-shim cinterop metadata for package $targetPackageName in ${unpackedKlibDir.absolutePath}"
            }

            targetFragments.forEach(::annotateFragment)

            writeLinkdata(
                linkdataDir = linkdataDir,
                metadataProvider = metadataProvider,
                serializedMetadata = metadata.write(),
            )
            patchManifest(archiveDir.resolve("manifest"))
            repackKlib(
                unpackedKlibInputDir = tempKlibDir,
                archiveDir = archiveDir,
                packedKlibFile = packedKlibFile,
            )
        } finally {
            tempKlibDir.deleteRecursively()
        }
    }

    private fun resolveArchiveDir(unpackedKlibDir: File): File = when {
        unpackedKlibDir.resolve("linkdata").isDirectory -> unpackedKlibDir
        unpackedKlibDir.resolve("default/linkdata").isDirectory -> unpackedKlibDir.resolve("default")
        else -> unpackedKlibDir
    }

    private fun annotateFragment(fragment: KmModuleFragment) {
        val packageFragment = requireNotNull(fragment.pkg) {
            "Expected package metadata for ${fragmentPackageName(fragment)}"
        }
        packageFragment.functions.forEach(::annotate)
        packageFragment.properties.forEach(::annotate)
        packageFragment.typeAliases.forEach(::annotate)
        fragment.classes.forEach(::annotate)
    }

    private fun annotate(clazz: KmClass) {
        clazz.annotations.addInternalNativeRpcApiMarker()
        clazz.constructors.forEach(::annotate)
        clazz.functions.forEach(::annotate)
        clazz.properties.forEach(::annotate)
        clazz.typeAliases.forEach(::annotate)
    }

    private fun annotate(constructor: KmConstructor) {
        constructor.annotations.addInternalNativeRpcApiMarker()
    }

    private fun annotate(function: KmFunction) {
        function.annotations.addInternalNativeRpcApiMarker()
    }

    private fun annotate(property: KmProperty) {
        property.annotations.addInternalNativeRpcApiMarker()
    }

    private fun annotate(typeAlias: KmTypeAlias) {
        typeAlias.annotations.addInternalNativeRpcApiMarker()
    }

    private fun MutableList<KmAnnotation>.addInternalNativeRpcApiMarker() {
        if (none { it.className == internalNativeRpcApiAnnotationClassName }) {
            add(internalNativeRpcApiAnnotation)
        }
    }

    private fun fragmentPackageName(fragment: KmModuleFragment): String {
        return fragmentPackageNameAccessor.invoke(null, fragment) as String
    }

    private fun writeLinkdata(
        linkdataDir: File,
        metadataProvider: LinkdataLibraryProvider,
        serializedMetadata: KlibModuleMetadata.SerializedKlibMetadata,
    ) {
        linkdataDir.resolve("module").writeBytes(serializedMetadata.header)

        val serializedByPackage = serializedMetadata.fragmentNames.zip(serializedMetadata.fragments)

        serializedByPackage.forEach { (packageName, fragments) ->
            val packageDir = linkdataDir.resolve(metadataProvider.packageDirectoryName(packageName))
            val outputNames = metadataProvider.existingPartNames(packageName).takeIf { it.size == fragments.size }
                ?: generatedPartNames(packageName, fragments.size)

            packageDir.deleteRecursively()
            packageDir.mkdirs()

            fragments.forEachIndexed { index, bytes ->
                packageDir.resolve(outputNames[index]).writeBytes(bytes)
            }
        }
    }

    private fun generatedPartNames(packageName: String, fragmentCount: Int): List<String> {
        val suffix = packageName.substringAfterLast('.', packageName).ifBlank { "root" }.replace('.', '_')
        return List(fragmentCount) { index -> "${index}_$suffix.knm" }
    }

    private fun patchManifest(manifestFile: File) {
        check(manifestFile.isFile) {
            "Missing KLIB manifest: ${manifestFile.absolutePath}"
        }
        val lines = manifestFile.readLines()
        val updatedLines = lines.map { line ->
            if (!line.startsWith("depends=")) return@map line

            val dependencies = line.removePrefix("depends=")
                .split(' ')
                .filter(String::isNotBlank)
                .toMutableList()
            if (internalNativeRpcApiDependencyUniqueName !in dependencies) {
                dependencies += internalNativeRpcApiDependencyUniqueName
            }
            "depends=${dependencies.joinToString(" ")}"
        }
        manifestFile.writeText(updatedLines.joinToString(separator = "\n", postfix = "\n"))
    }

    private fun repackKlib(unpackedKlibInputDir: File, archiveDir: File, packedKlibFile: File) {
        val archiveRoot = if (archiveDir.parentFile == unpackedKlibInputDir) {
            unpackedKlibInputDir
        } else {
            archiveDir.parentFile ?: error("Expected ${archiveDir.absolutePath} to have a parent directory")
        }

        packedKlibFile.parentFile.mkdirs()
        ZipOutputStream(packedKlibFile.outputStream().buffered()).use { zip ->
            Files.walk(archiveDir.toPath()).use { paths ->
                paths.filter { path -> Files.isRegularFile(path) }
                    .sorted()
                    .forEach { path ->
                        val entryName = archiveRoot.toPath()
                            .relativize(path)
                            .toString()
                            .replace(File.separatorChar, '/')
                        zip.putNextEntry(ZipEntry(entryName))
                        Files.copy(path, zip)
                        zip.closeEntry()
                    }
            }
        }
    }

    private class LinkdataLibraryProvider(
        private val linkdataDir: File,
    ) : KlibModuleMetadata.MetadataLibraryProvider {
        override val moduleHeaderData: ByteArray
            get() = linkdataDir.resolve("module").readBytes()

        override fun packageMetadataParts(fqName: String): Set<String> =
            packagePartFiles(fqName).mapTo(linkedSetOf()) { it.name }

        override fun packageMetadata(fqName: String, partName: String): ByteArray =
            linkdataDir.resolve(packageDirectoryName(fqName)).resolve(partName).readBytes()

        fun existingPartNames(packageName: String): List<String> = packagePartFiles(packageName).map(File::getName)

        fun packageDirectoryName(packageName: String): String = "package_" + packageName.replace('.', '_')

        private fun packagePartFiles(packageName: String): List<File> {
            val packageDir = linkdataDir.resolve(packageDirectoryName(packageName))
            check(packageDir.isDirectory) {
                "Missing package metadata directory for $packageName: ${packageDir.absolutePath}"
            }
            return packageDir.listFiles()
                .orEmpty()
                .filter(File::isFile)
                .sortedBy(File::getName)
        }
    }
}
